OpenRA.Mods.D2k/Traits/SpiceBloom.cs

```diff
@@ -91,45 +103,50 @@ public void Tick(Actor self)
 
 		public void Killed(Actor self, AttackInfo e)
 		{
-			var pieces = self.World.SharedRandom.Next(info.Pieces[0], info.Pieces[1]) * ticks / growTicks;
-			if (pieces < info.Pieces[0])
-				pieces = info.Pieces[0];
-
-			for (var i = 0; pieces > i; i++)
+			if (!string.IsNullOrEmpty(info.Weapon))
 			{
-				var cells = OpenRA.Traits.Util.RandomWalk(self.Location, self.World.SharedRandom);
-				var cell = cells.Take(info.Range).SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p)).Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
-				if (cell == null)
-					cell = cells.Take(info.Range).Random(self.World.SharedRandom);
+				var pieces = self.World.SharedRandom.Next(info.Pieces[0], info.Pieces[1]) * ticks / growTicks;
+				if (pieces < info.Pieces[0])
+					pieces = info.Pieces[0];
 
-				var args = new ProjectileArgs
+				var cells = self.World.Map.FindTilesInAnnulus(self.Location, 1, info.Range)
+					.SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p));
+
+				for (var i = 0; pieces > i; i++)
 				{
-					Weapon = self.World.Map.Rules.Weapons[info.Weapon.ToLowerInvariant()],
-					Facing = 0,
+					var cell = cells.Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
+					if (cell == null)
+						cell = cells.Random(self.World.SharedRandom);
```

Comment:
Looks like the old code was attempting to take a random cell in range but skipping over anything already full of resource. If there were no such cells (everything was already maxed) it just used any random cell in range instead.

Try moving the `SkipWhile` currently on the `cells` variable back onto the first random search line - that ought to make it work like the original code again as best I can tell.